### ISSUE TITLE
aws-cdk-lib: Fargate task check requires at least 21G of storage but docs say 20G is the mininum

### ISSUE THREAD

**Issue by JoshMcCullough** [aws/aws-cdk#31291]
### Describe the bug

When configuring a `FargateTaskDefinition`, passing `ephemeralStorageGiB: 20` throws the following error, even though [the docs suggest that 20G is the minimum allowed](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/fargate-task-storage.html#:~:text=Fargate%20Windows%20container%20platform%20versions,-Version%201.0.&text=platform%20version%201.0.-,0%20or%20later%20receive%20a%20minimum%20of%2020%20GiB%20of,parameter%20in%20your%20task%20definition.).

> Ephemeral storage size must be between 21GiB and 200GiB

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

`ephemeralStorageGiB: 20` should be allowed, or the docs should be updated.

### Current Behavior

`ephemeralStorageGiB: 20` is not allowed, minimum value must be `21`.

### Reproduction Steps

1. Configure a `FargateTaskDefinition`.
2. Set: `ephemeralStorageGiB: 20`
3. Try to `synth`.

### Possible Solution

Fix `aws-cdk-lib/aws-ecs/lib/fargate/fargate-task-definition.js` to check `(props.ephemeralStorageGiB<20||props.ephemeralStorageGiB>200), or fix the docs.

### CDK CLI Version

2.155.0

### Node.js Version

18.19.0

### OS

Manjaro Linux

### Language

TypeScript

### Language Version

5.5.4

**Comment by ashishdhingra** [comment]
The error is thrown by the validation check [here](https://github.com/aws/aws-cdk/blob/a6c2ad6cbd384be6e015245b95b4603afea91902/packages/aws-cdk-lib/aws-ecs/lib/fargate/fargate-task-definition.ts#L169).

Per [AWS::ECS::TaskDefinition EphemeralStorage](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-ephemeralstorage.html), `SizeInGiB` has the minimum supported value of `20 GiB` and the maximum supported value of `200 GiB`.

**Comment by mazyu36** [comment]
Hello,

I modified the if statement and attempted to deploy with 20GB, but I encountered the following error. 
The current implementation seems correct.

![スクリーンショット 2024-09-03 22 23 43](https://github.com/user-attachments/assets/ab51d9fd-9298-4767-8127-7eca92159d21)

It is also noted in the Management Console.
![image](https://github.com/user-attachments/assets/a6fef0e3-74c6-4bce-8b49-0f7629dd2d0f)

**Comment by ashishdhingra** [comment]
> Hello.
> 
> I modified the if statement and attempted to deploy with 20GB, but I encountered the following error. The current implementation seems correct.
> 
> ![スクリーンショット 2024-09-03 22 23 43](https://private-user-images.githubusercontent.com/13781813/364378540-ab51d9fd-9298-4767-8127-7eca92159d21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0Njk2ODMsIm5iZiI6MTcyNTQ2OTM4MywicGF0aCI6Ii8xMzc4MTgxMy8zNjQzNzg1NDAtYWI1MWQ5ZmQtOTI5OC00NzY3LTgxMjctN2VjYTkyMTU5ZDIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA0VDE3MDMwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmZjg1ZmRkZTE2YjI0YTIwZTQ4ODQ2OTU3NDIyYTlhYjUxNDE3MTAwOWU1OGQ3MDYzZmMzNjgzYzgxOThhZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UJlrOex41djaoFa0jhN9qUgylmftoo4kwZ0Kz9gtfwY)
> 
> It is also noted in the Management Console. ![image](https://private-user-images.githubusercontent.com/13781813/364379098-a6fef0e3-74c6-4bce-8b49-0f7629dd2d0f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0Njk2ODMsIm5iZiI6MTcyNTQ2OTM4MywicGF0aCI6Ii8xMzc4MTgxMy8zNjQzNzkwOTgtYTZmZWYwZTMtNzRjNi00YmNlLThiNDktMGY3NjI5ZGQyZDBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA0VDE3MDMwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyNmU2MTY1MDM4MGU5MTJhMTdmY2RlYmE3MmE0MmZiMzg2ODQ4ZTIzOWEwY2RkMmYzNGVjZDFmZjMzMTRkYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5JWZKanvvo4rebs6OxYuDuvpyz0771jnp6Zo6fkEa4s)

@mazyu36 Good observation. I will revalidate and would probably open ticket with CloudFormation team for documentation correction.

**Comment by ashishdhingra** [comment]
Tried to bypass CDK validation using below code (uses escape hatch):
```TypeScript
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as cdk from 'aws-cdk-lib';
import * as ecs from 'aws-cdk-lib/aws-ecs';

import path = require('path');


export class CdktestStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create an ECS cluster
    const vpc = ec2.Vpc.fromLookup(this, 'MyDefaultVpc', {
      isDefault: true
    });
    const cluster = new ecs.Cluster(this, 'Cluster', { vpc });

    // Add capacity to it
    cluster.addCapacity('DefaultAutoScalingGroupCapacity', {
      instanceType: new ec2.InstanceType("t2.xlarge"),
      desiredCapacity: 3,
    });

    const taskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDef', { ephemeralStorageGiB: 21 });
    (taskDefinition.node.defaultChild as ecs.CfnTaskDefinition).addPropertyOverride('EphemeralStorage', 20);

    taskDefinition.addContainer('DefaultContainer', {
      image: ecs.ContainerImage.fromRegistry("amazon/amazon-ecs-sample"),
      memoryLimitMiB: 512,
    });

    // Instantiate an Amazon ECS Service
    const ecsService = new ecs.FargateService(this, 'FargateService', {
      cluster,
      taskDefinition,
    });
  }
}
```
Although it is able to set `EphemeralStorage` to `20` in CloudFormation template:
```YAML
Resources:
  ...
  ...
  TaskDef54694570:
    Type: AWS::ECS::TaskDefinition
    Properties:
      ContainerDefinitions:
        - Essential: true
          Image: amazon/amazon-ecs-sample
          Memory: 512
          Name: DefaultContainer
      Cpu: "256"
      EphemeralStorage: 20
      Family: CdktestStackTaskDefB8572B74
      Memory: "512"
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      TaskRoleArn:
        Fn::GetAtt:
          - TaskDefTaskRole1EDB4A67
          - Arn
    Metadata:
      aws:cdk:path: CdktestStack/TaskDef/Resource
...
```
Running `cdk deploy` fails with the below error:
```
CdktestStack: deploying... [1/1]
CdktestStack: creating CloudFormation changeset...
11:43:56 AM | CREATE_FAILED        | AWS::ECS::TaskDefinition           | TaskDef54694570
Resource handler returned message: "Invalid request provided: Create TaskDefinition: EphemeralStorage size should be at least 21 (Service: AmazonECS; Status Code: 400; Erro
r Code: ClientException; Request ID: d75746d2-7eac-4c17-8114-56c25b9e8c14; Proxy: null)" (RequestToken: 45f51668-4303-c8cf-2610-189ffcbb2abc, HandlerErrorCode: InvalidReque
st)


 ❌  CdktestStack failed: Error: The stack named CdktestStack failed creation, it may need to be manually deleted from the AWS console: ROLLBACK_COMPLETE: Resource handler returned message: "Invalid request provided: Create TaskDefinition: EphemeralStorage size should be at least 21 (Service: AmazonECS; Status Code: 400; Error Code: ClientException; Request ID: d75746d2-7eac-4c17-8114-56c25b9e8c14; Proxy: null)" (RequestToken: 45f51668-4303-c8cf-2610-189ffcbb2abc, HandlerErrorCode: InvalidRequest)
    at FullCloudFormationDeployment.monitorDeployment (/opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:452:10567)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Object.deployStack2 [as deployStack] (/opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:455:200276)
    at async /opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:455:181698

 ❌ Deployment failed: Error: The stack named CdktestStack failed creation, it may need to be manually deleted from the AWS console: ROLLBACK_COMPLETE: Resource handler returned message: "Invalid request provided: Create TaskDefinition: EphemeralStorage size should be at least 21 (Service: AmazonECS; Status Code: 400; Error Code: ClientException; Request ID: d75746d2-7eac-4c17-8114-56c25b9e8c14; Proxy: null)" (RequestToken: 45f51668-4303-c8cf-2610-189ffcbb2abc, HandlerErrorCode: InvalidRequest)
    at FullCloudFormationDeployment.monitorDeployment (/opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:452:10567)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Object.deployStack2 [as deployStack] (/opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:455:200276)
    at async /opt/homebrew/lib/node_modules/aws-cdk/lib/index.js:455:181698
cdk****************************************************
```

So the existing validation is correct. We would need to cut ticket to CloudFormation team to get the documentation corrected.

**Comment by ashishdhingra** [comment]
@JoshMcCullough Looking at the AWS console screenshot, I now understand what is happening.

<img width="1625" alt="Screenshot 2024-09-06 at 12 28 59 PM" src="https://github.com/user-attachments/assets/5cf3056d-de1c-4e7e-b478-e4ed0aa34537">

- It specifies that `By default, your tasks hosted on AWS Fargate receive a minimum of 20 GiB of ephemeral storage.`.
- However, `To specify a custom amount of ephemeral storage, specify a value between 21 GiB up to a maximum of 200 GiB.`

In other words, when trying to explicitly specify `EphemeralStorage` in CloudFormation template, one needs to specify minimum value of `21 GiB`. Specifying custom value as `20 GiB` doesn't make sense since it would anyway default to this value if no value is specified.

Based on above understanding, the validation in CDK and the default value mentioned in documentation is correct. Please review and confirm if this issue could be closed.

Thanks,
Ashish

**Comment by JoshMcCullough** [comment]
Okay. It's strange that 20 is technically valid but only if you don't specify any value.

**Comment by ashishdhingra** [comment]
> Okay. It's strange that 20 is technically valid but only if you don't specify any value.

@JoshMcCullough Yes, it's really strange and was confusing in the beginning until I noticed a note in AWS console. Closing this ticket for now. Feel free to open another ticket in case you see other issues.
